### PR TITLE
Membership benchmarking

### PR DESCRIPTION
**Pull request by iorveth**
The scope of this PR includes:
1. Benchmarks for membership extrinsics.
2. Generated weight functions.
3. Extrinsics weight annotations.
4. Newly introduced `MembershipWorkingGroupHelper` trait and respective logic to be able to set up lead for membership benchmarking purposes.
5. Upgraded mock runtimes for pallets dependent on membership in dev enviroment.

### CODE REVIEW

File: runtime-modules/membership/src/lib.rs [hunk 1]
```diff
@@ -313,7 +346,28 @@ decl_module! {
...
+        ///
+        /// <weight>
+        ///
+        /// ## Weight
+        /// `O (W)` where:
+        /// - `W` is the handle length
+        /// - DB:
+        ///    - O(W)
+        /// # </weight>
+        #[weight = if params.referrer_id.is_some() {
```
Comment (conectado) [inline runtime-modules/membership/src/lib.rs]: I think it'd be better to move the weight calculation to a separate function

File: runtime-modules/membership/src/benchmarking.rs [hunk 2]
```diff
@@ -0,0 +1,775 @@
...
+        let second_member_id = member_id + T::MemberId::one();
+
+        assert_eq!(MemberIdByHandleHash::<T>::get(second_handle_hash), second_member_id);
+
+        assert_eq!(MembershipById::<T>::get(second_member_id), membership);
+
+        assert_last_event::<T>(RawEvent::MemberRegistered(second_member_id).into());
+    }
+
+    update_profile{
```
Comment (conectado) [inline runtime-modules/membership/src/benchmarking.rs:218]: Please test for all parameters possible length(name, handle, avatar_uri, about)

File: runtime-modules/membership/src/lib.rs [hunk 3]
```diff
@@ -412,7 +479,22 @@ decl_module! {
...
+        /// <weight>
+        ///
+        /// <weight>
+        ///
+        /// ## Weight
+        /// `O (1)`
+        /// - DB:
+        ///    - O(1) doesn't depend on the state or parameters
+        /// # </weight>
+        #[weight = match (new_root_account.is_some(), new_controller_account.is_some()) {
```
Comment (conectado) [inline runtime-modules/membership/src/lib.rs]: Please move weight calculation to a separate function

File: runtime-modules/membership/src/benchmarking.rs [hunk 4]
```diff
@@ -0,0 +1,775 @@
...
+
+        assert_eq!(MembershipById::<T>::get(first_member_id), first_membership);
+
+        assert_eq!(MembershipById::<T>::get(second_member_id), second_membership);
+
+        assert_last_event::<T>(RawEvent::InvitesTransferred(first_member_id, second_member_id, number_of_invites).into());
+    }
+
+    invite_member {
+        let member_id = 0;
```
Comment (conectado) [inline runtime-modules/membership/src/benchmarking.rs:423]: Please benchmark with all different parameters length in `InviteMembershipParameters`

**Review by conectado — CHANGES_REQUESTED** [review]
Great work!

Just a couple of comments. 

But more importantly, we could do without introducing the trait in general, we can create a trait only within the benchmarking features. [See this example(the tests aren't fixed here)](https://github.com/conectado/joystream/commit/271a6b7df4c37b472fffe5283f7e1d556286fb6b).

File: runtime-modules/membership/src/benchmarking.rs (hunk 2 above)
Comment (iorveth) [reply runtime-modules/membership/src/benchmarking.rs:218]: It seems, that they do not affect benchmarking weight at all.
`	// WARNING! Some components were not used: ["j"]`
Only `handle` possible lengths affects benchmarking. 
We do not perform any computations/db operations with any of those parameters.

File: runtime-modules/membership/src/benchmarking.rs (hunk 4 above)
Comment (iorveth) [reply runtime-modules/membership/src/benchmarking.rs:423]: It seems, that they do not affect benchmarking weight at all.
`// WARNING! Some components were not used: ["j"]`

Only `handle` possible lengths affects benchmarking.
We do not perform any computations/db operations with any of those parameters.

File: runtime-modules/membership/src/lib.rs (hunk 3 above)
Comment (iorveth) [reply runtime-modules/membership/src/lib.rs]: Done https://github.com/Joystream/joystream/pull/2024/commits/187fa9e4ba17abd5e1c97dd2fc8db253f282dd7f

File: runtime-modules/membership/src/lib.rs (hunk 1 above)
Comment (iorveth) [reply runtime-modules/membership/src/lib.rs]: Done https://github.com/Joystream/joystream/pull/2024/commits/187fa9e4ba17abd5e1c97dd2fc8db253f282dd7f

**Comment by iorveth** [comment]
`But more importantly, we could do without introducing the trait in general, we can create a trait only within the benchmarking features. See this example(the tests aren't fixed here).
`

Thank you for the helpful suggestion!

File: runtime-modules/membership/src/benchmarking.rs (hunk 2 above)
Comment (conectado) [reply runtime-modules/membership/src/benchmarking.rs:218]: I think it's better if we leave the benchmarking for those parameters on and we manually modify the trait(this won't be neccesary with the next Substrate update). It's how we have handled it in the other pallets and sometimes unexpectedly these parameters have some footprint in processing time.

File: runtime-modules/membership/src/benchmarking.rs (hunk 4 above)
Comment (conectado) [reply runtime-modules/membership/src/benchmarking.rs:423]: I think it's better if we leave the benchmarking for those parameters on and we manually modify the trait(this won't be neccesary with the next Substrate update). It's how we have handle it in the other pallets and sometimes unexpectedly these parameters have some footprint in processing time.

**Review by conectado — CHANGES_REQUESTED** [review]
I just recommend that we would benchmark unused parameters even if they don't affect the results.

**Pull request merged**